metaflow/datatools/s3op.py

```diff
@@ -106,66 +157,125 @@ def worker(queue, mode):
                             raise RuntimeError("Could not load file")
                     tmp.close()
                     os.rename(tmp.name, url.local)
+                except ClientError as err:
+                    error_code = normalize_client_error(err)
+                    if error_code == 404:
+                        pass # We skip this
+                    else:
+                        raise
                 except:
                     # TODO specific error message for out of disk space
                     tmp.close()
                     os.unlink(tmp.name)
                     raise
                 # If we have metadata that we retrieved, we also write it out
                 # to a file
-                if url.content_type or url.metadata:
+                if result_info:
                     with open('%s_meta' % url.local, mode='w') as f:
-                        args = {}
-                        if url.content_type:
-                            args['content_type'] = url.content_type
-                        if url.metadata is not None:
-                            args['metadata'] = url.metadata
+                        args = {'size': result_info['size']}
+                        if result_info['content_type']:
+                            args['content_type'] = result_info['content_type']
+                        if result_info['metadata'] is not None:
+                            args['metadata'] = result_info['metadata']
                         json.dump(args, f)
+                    # Finally, we push out the size to the result_pipe since
+                    # the size is used for verification and other purposes and
+                    # we want to avoid file operations for this simple process
+                    result_pipe.send((idx, result_info['size']))
```

Comment:
this call will block if the pipe buffer is full. It can get full looking at the receiving side below.

Since you open the `_meta` file above anyways, why not write these results in the `_meta` file, which is guaranteed to be non_blocking and get rid of the pipe?

Reply:
So yes the pipe could block. It's more OS dependent but on Linux, typically, that would take a few thousand files to block (pipes are typically 64K it seems) so possible but not highly likely. To get to the second question, I could indeed write the information in the file (it actually is contained there). The reason I was also pushing it to a pipe is because of the fact that the size is used quickly for lots of things (verifying existence being one of them). I thought that using files for this might be a bit more expensive). For a few files, it wouldn't be a big problem (but then in that case, pipes would not block either) and for many files where pipes could be problematic, opening and closing thousands of small files might also not be ideal (but I admittedly did not benchmark it). Ideally it would also be a separate file (to keep the index -- yes, we could map but that also seems more costly). We can discuss a bit more though because this is by no means the ideal solution.

Reply:
yeah, a concern about many small files is a valid one. I assume you are mainly worried about the read side, since it seems you are writing a separate file for each object above already.

An interesting alternative might be to *append* all results in a single file. Small appends are atomic https://stackoverflow.com/questions/1154446/is-file-append-atomic-in-unix/24270790#24270790

Reply:
Addressed in the refactoring using a single file per worker.